lib/mongo/crypt/binary.rb

```diff
@@ -50,6 +63,60 @@ def initialize(data=nil)
         end
       end
 
+      # Initialize a Binary object from an existing pointer to a mongocrypt_binary_t
+      # object.
+      #
+      # @param [ FFI::Pointer ] pointer A pointer to an existing
+      #   mongocrypt_binary_t object
+      #
+      # @return [ Mongo::Crypt::Binary ] A new binary object
+      def self.from_pointer(pointer)
+        self.new(pointer: pointer)
+      end
+
+      # Initialize a Binary object with a string. The Binary object will store a
+      # copy of the specified string and destroy the allocated memory when
+      # it goes out of scope.
+      #
+      # @param [ String ] data A string to be wrapped by the Binary object
+      #
+      # @return [ Mongo::Crypt::Binary ] A new binary object
+      def self.from_data(data)
+        self.new(data: data)
+      end
+
+      # Overwrite the existing data wrapped by this Binary object
+      #
+      # @note The data passed in must not take up more memory than the
+      # original memory allocated to the underlying mongocrypt_binary_t
+      # object. Do NOT use this method unless required to do so by libmongocrypt.
+      #
+      # @param [ String ] data The new string data to be wrapped by this binary object
+      #
+      # @return [ true | false ] Whether the operation was successful
+      def write(data)
+        # Cannot write a string that's longer than the space currently allocated
+        # by the mongocrypt_binary_t object
+        return false if Binding.mongocrypt_binary_len(@bin) < data.length
+
+        bytes = data.unpack('C*')
+
+        begin
+          @data_p.clear if @data_p
+
+          # # FFI::MemoryPointer automatically frees memory when it goes out of scope
+          @data_p = FFI::MemoryPointer.new(bytes.length)
+                      .write_array_of_uint8(bytes)
+
+          Binding.mongocrypt_binary_data(@bin)
+                    .write_array_of_uint8(bytes)
+        rescue FFI::NullPointerError
+          return false
```

Comment:
When would this case happen?

Reply:
If you initialize a mongocrypt_binary_t object with no data and then try to write data to it. I'll write a comment.

Reply:
I've realized this isn't necessary because of the conditional above it. I'll remove this case.

Reply:
Sounds good - I would think that it is a bug in the driver to make libmongocrypt reference null pointers (for practical reasons like it interferes with debugging, even if LMC could/should provide better diagnostics in such cases).